File: Solutions/K-diff_Pairs_in_an_Array/K-diff_Pairs_in_an_Array.cs
public class Solution {
    public int FindPairs(int[] nums, int k) {
        if (k == 0) {
            Dictionary<int,int> dict = new Dictionary<int,int>();
            foreach (int i in nums ) {
                if (dict.ContainsKey(i))
                    dict[i] += 1;
                else
                    dict[i] = 1;
            }
            return dict.Values.Where(c => c >= 2).Count();
        }

        int result = 0;
        HashSet<int> hset = new HashSet<int>();
        foreach (int i in nums) {
            if (!hset.Contains(i)) {
                if (hset.Contains(i+k))
                    result++;
                if (hset.Contains(i-k))
                    result++;
            }
            hset.Add(i);
        }
        return result;
    }
}

//Most common
public class Solution {
    public int FindPairs(int[] nums, int k) {
        
            int ans = 0;
            Dictionary<int, int> mm = new Dictionary<int, int>();

            for (int i = 0; i < nums.Length; i++)
            {
                if (mm.ContainsKey(nums[i]))
                {
                    mm[nums[i]]++;
                }
                else 
                {
                    mm.Add(nums[i], 1);
                }
            }

            for (int i = 0; i < nums.Length; i++)
            {
                if (k > 0 && mm.ContainsKey(nums[i] + k) && mm[nums[i]] > 0)
                {
                    mm[nums[i]] = -1;
                    ans++;
                }

                if (k == 0 && mm.ContainsKey(nums[i]) && mm[nums[i]] > 1)
                {
                    ans++;
                    mm[nums[i]] = -1;
                }
            }

            return ans;
    }
}

//Fastest
public class Solution {
    public int FindPairs(int[] nums, int k) {
            if (k < 0)
            {
                return 0;
            }

            int answer = 0;
            Dictionary<int, int> dic = new Dictionary<int, int>();
            foreach (int n in nums)
            {

                if (dic.ContainsKey(n))
                {
                    if (dic[n] <= 1 &&  k==0) 
                    {
                        answer++;
                        dic[n]++;
                    }
                    
                    continue;
                }

                if (dic.ContainsKey(n - k)) 
                {
                    answer++;
                    dic[n - k]++;
                }
                if (dic.ContainsKey(n + k))
                {
                    answer++;
                    dic[n + k]++;
                }

                dic.Add(n, 1);
        
            }

            return answer;        
    }
}

File: Solutions/K-diff_Pairs_in_an_Array/K-diff_Pairs_in_an_Array.go
//Most common
import "sort"

func findPairs(nums []int, k int) int {
    if k < 0 {
        return 0
    }
    sort.Ints(nums)
    modified := make([]int,len(nums))
    for i := range nums {
        modified[i] = nums[i] - k
    }
    i, j := 0, 0
    count := 0
    used := make(map[int]bool)
    for i < len(modified) && j < len(nums) {
        if modified[i] == nums[j] && i != j {
            if _, ok := used[modified[i] + nums[j]]; !ok {
                count++
                used[modified[i] + nums[j]] = true
            }
        }
        if modified[i] < nums[j] {
            i++
        } else {
            j++
        }
    }
    return count
}

//Second fastest
func findPairs(nums []int, k int) int {
    if k < 0 || len(nums) <= 0 {
        return 0
    }
    
    m := make(map[int]int)
    for _, v := range nums {
        m[v]++
    }
    
    result := 0
    for key, _ := range m {
        if k == 0 && m[key] > 1 {
            result++
        }
        if k > 0 && m[key+k] > 0 {
            result++
        }
    }
    return result
}

//Fastest
func findPairs(nums []int, k int) int {
	sort.Ints(nums)

	m := make(map[string]bool)
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[j]-nums[i] == k {
				m[fmt.Sprintf("%d%d", nums[j], nums[i])] = true
			}

			if nums[j]-nums[i] > k {
				break
			}
		}
	}

	return len(m)
}

//Another fastest
func findPairs(nums []int, k int) int {
        
	var res int
	m := map[int]int{}
	
	for _, n := range nums {
			m[n]++
	}
	
	for n, count := range m {
			if k > 0 && m[k+n] != 0 {
					res++
			} else if k == 0 && count > 1 {
					res++
			}
	}
	
	return res
}